### PR TITLE
feat:  Tokenfactory SetMetadata

### PR DESCRIPTION
**Pull request by sunnya97**
## What is the purpose of the change

This PR allows the admin of a tokenfactory denom to set its bank.Metadata, thereby changing its display units, name, and symbol.

## Brief Changelog

  - *The metadata is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *Daemons retrieve the RPC data from the blob cache*

## Testing and Verifying

This change added unit tests in x/tokenfactory/keeper/admins_test.go.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? yes
  - How is the feature or change documented? x/tokenfactory/README

### CODE REVIEW

File: x/tokenfactory/keeper/msg_server.go [hunk 1]
```diff
@@ -158,3 +159,39 @@ func (server msgServer) ChangeAdmin(goCtx context.Context, msg *types.MsgChangeA
...
+func (server msgServer) SetDenomMetadata(goCtx context.Context, msg *types.MsgSetDenomMetadata) (*types.MsgSetDenomMetadataResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Defense in depth validation of metadata
+	err := msg.Metadata.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println(msg.Metadata.Base)
```
Comment (sunnya97) [inline x/tokenfactory/keeper/msg_server.go:172]:
```suggestion
```

File: x/tokenfactory/keeper/msg_server.go [hunk 2]
```diff
@@ -158,3 +159,37 @@ func (server msgServer) ChangeAdmin(goCtx context.Context, msg *types.MsgChangeA
...
+		return nil, err
+	}
+
+	authorityMetadata, err := server.Keeper.GetAuthorityMetadata(ctx, msg.Metadata.Base)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println(msg.Sender)
+	fmt.Println(authorityMetadata)
```
Comment (sunnya97) [inline x/tokenfactory/keeper/msg_server.go:178]:
```suggestion
```

File: x/tokenfactory/README.md [hunk 3]
```diff
@@ -99,6 +99,18 @@ message MsgChangeAdmin {
 }
 ```
 
+### SetDenomMetadata
+
+Setting of metadata for a specific denom is only allowed for the admin of the denom.
+It allows the overwriting of the denom metadata in the bank module.
+
+```go
+message MsgChangeAdmin {
```
Comment (mattverse) [inline x/tokenfactory/README.md:108]:
```suggestion
message MsgSetDenomMetadata {
```

File: x/tokenfactory/keeper/msg_server.go [hunk 4]
```diff
@@ -158,3 +158,34 @@ func (server msgServer) ChangeAdmin(goCtx context.Context, msg *types.MsgChangeA
...
+
+	if msg.Sender != authorityMetadata.GetAdmin() {
+		return nil, types.ErrUnauthorized
+	}
+
+	server.Keeper.bankKeeper.SetDenomMetaData(ctx, msg.Metadata)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.TypeMsgChangeAdmin,
```
Comment (mattverse) [inline x/tokenfactory/keeper/msg_server.go:184]:
```suggestion
			types.TypeMsgSetDenomMetadata,
```

**Review by mattverse — APPROVED** [review]
LGTM! 

Can we register codec for this as well before merging this?

**Comment by ValarDragon** [comment]
Thus conflicts with the goal of the state machine having a reverse map from human readable name to bank denom

**Comment by sunnya97** [comment]
I think the metadata stored in the bank module should a token's self-declared name/metadata (similar to how all ERC20s can self-declare their metadata).  Then, the state machine can have its own on-chain curated assetlist, separate from the self-declared info).

File: x/tokenfactory/keeper/admins_test.go [hunk 5]
```diff
@@ -267,3 +268,131 @@ func (suite *KeeperTestSuite) TestChangeAdminDenom() {
...
+						Denom:    "uosmo",
+						Exponent: 6,
+					},
+				},
+				Base:    fmt.Sprintf("factory/%s/litecoin", suite.TestAccs[0].String()),
+				Display: "uosmo",
+				Name:    "OSMO",
+				Symbol:  "OSMO",
+			}),
+			expectedPass: false,
```
Comment (p0mvn) [inline x/tokenfactory/keeper/admins_test.go:323]: it's false by default so can be omitted
```suggestion
```

File: x/tokenfactory/keeper/admins_test.go [hunk 6]
```diff
@@ -267,3 +268,131 @@ func (suite *KeeperTestSuite) TestChangeAdminDenom() {
...
+						Denom:    "uosmoo",
+						Exponent: 6,
+					},
+				},
+				Base:    "uosmo",
+				Display: "uosmoo",
+				Name:    "OSMO",
+				Symbol:  "OSMO",
+			}),
+			expectedPass: false,
```
Comment (p0mvn) [inline x/tokenfactory/keeper/admins_test.go:344]:
```suggestion
```

File: x/tokenfactory/keeper/admins_test.go [hunk 7]
```diff
@@ -267,3 +268,131 @@ func (suite *KeeperTestSuite) TestChangeAdminDenom() {
...
+				Base:    suite.defaultDenom,
+				Display: "uosmo",
+				Name:    "OSMO",
+				Symbol:  "OSMO",
+			}),
+			expectedPass: false,
+		},
+	} {
+		suite.Run(fmt.Sprintf("Case %s", tc.desc), func() {
+			_, err := suite.msgServer.SetDenomMetadata(sdk.WrapSDKContext(suite.Ctx), &tc.msgSetDenomMetadata)
```
Comment (p0mvn) [inline x/tokenfactory/keeper/admins_test.go:386]: nit: we can also make an assertion that the response is not nil

File: x/tokenfactory/types/msgs.go [hunk 8]
```diff
@@ -202,3 +204,43 @@ func (m MsgChangeAdmin) GetSigners() []sdk.AccAddress {
...
+
+	return nil
+}
+
+func (m MsgSetDenomMetadata) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&m))
+}
+
+func (m MsgSetDenomMetadata) GetSigners() []sdk.AccAddress {
+	sender, _ := sdk.AccAddressFromBech32(m.Sender)
```
Comment (p0mvn) [inline x/tokenfactory/types/msgs.go:244]: Can error be propagated to the caller?

**Review by p0mvn — APPROVED** [review]

File: x/tokenfactory/types/msgs.go (hunk 8 above)
Comment (sunnya97) [reply x/tokenfactory/types/msgs.go:244]: I don't think so.  This is part of the sdk.Msg interface, so we can't change the function signature

File: x/tokenfactory/keeper/admins_test.go (hunk 5 above)
Comment (sunnya97) [reply x/tokenfactory/keeper/admins_test.go:323]: I think I prefer to still see visually when reading the code that this test case is expected to fail.

File: x/tokenfactory/keeper/admins_test.go (hunk 7 above)
Comment (sunnya97) [reply x/tokenfactory/keeper/admins_test.go]: Done

**Comment by ValarDragon** [comment]
I'm pretty concerned about this, as metadata already has users with people expecting safety guarantees.

E.g. integration into ledger

**Comment by sunnya97** [comment]
integration into ledger doesn't use metadata right now?

**Comment by ValarDragon** [comment]
Oh I thought it read from bank metadata, its actually just hardcoded into ledger: https://github.com/cosmos/ledger-cosmos/blob/78a1a260ee66702cc21c81dbcbffef8ebe9647f5/app/src/coin.h

**Pull request merged**